### ISSUE TITLE
METAnalyzer tries to get TriggerResults while on a Path

### ISSUE THREAD

**Issue by Dr15Jones** [cms-sw/cmssw#15769]
The METAnalyzer tries to get a series of TriggerResults
http://cmslxr.fnal.gov/lxr/source/DQMOffline/JetMET/src/METAnalyzer.cc?v=CMSSW_8_1_X_2016-08-25-2300#0061
http://cmslxr.fnal.gov/lxr/source/DQMOffline/JetMET/src/METAnalyzer.cc?v=CMSSW_8_1_X_2016-08-25-2300#0160
http://cmslxr.fnal.gov/lxr/source/DQMOffline/JetMET/src/METAnalyzer.cc?v=CMSSW_8_1_X_2016-08-25-2300#0163

The problem is it lives on a Path not an EndPath and therefore the TriggerResults it request is not available. 
http://cmslxr.fnal.gov/lxr/source/DQMOffline/JetMET/python/metDQMConfig_cfi.py?v=CMSSW_8_1_X_2016-08-25-2300#0009

Previously, this was tolerated by the framework but with the upcoming changes for threading efficiency we now will require that if a module asks for a data product that the scheduling system can actually run it. For this case, since the `caloMetDQMAnalyzer` module is on a Path and TriggerResults are not made until all Paths have run, this creates an impossible to run schedule.

The METAnalyzer class should be changed to accept empty InputTags for the TriggerResults and then the configurations should disable the results that are not going to be needed.

**Comment by Dr15Jones** [comment]
urgent

**Comment by Dr15Jones** [comment]
assign dqm

**Comment by Dr15Jones** [comment]
The updated schedule inconsistency checker [ #15740 ] has found additional DQM related problems. They are all related to asking for TriggerResults from the Path and not EndPath. The two new error messages are

from workflows
 140.1,140.2,140.4,145.0
```
TriggerResults after DQMMessageLogger [path dqmoffline_step], DQMMessageLogger after CentralityDQM [path dqmoffline_step], ... [lots more modules on path dqmoffline_step] ..., dqmEnvHLT after higPhotonJetHLTOfflineSource [path dqmoffline_step], higPhotonJetHLTOfflineSource consumes TriggerResults
```

From workflows 
5.1, 5.2,5.3,5.4,5.5, 5.6, 5.7
135.1,135.2,135.3,135.4,135.5,135.6,135.7,135.9,135.11,135.12
```
TriggerResults after dqmInfoTauV [path dqmoffline_step], dqmInfoTauV after RunHPSValidationRealMuonsData [path dqmoffline_step], RunHPSValidationRealMuonsData consumes TriggerResults
```

**Comment by Dr15Jones** [comment]
From the 5.1 workflow configuration
```
>>> print process.RunHPSValidationRealMuonsData.dumpPython()
cms.EDAnalyzer("TauTagValidation",
    DataType = cms.string('Leptons'),
    ExtensionName = cms.string('RealMuonsData'),
    GenericTriggerSelection = cms.PSet(
        andOr = cms.bool(False),
        andOrHlt = cms.bool(True),
        dbLabel = cms.string('PFTauDQMTrigger'),
        errorReplyHlt = cms.bool(False),
        hltDBKey = cms.string('TauTriggerForALLMuDataset'),
        hltInputTag = cms.InputTag("TriggerResults","","HLT"),
        hltPaths = cms.vstring('HLT_IsoMu24_eta2p1_v*'),
        verbosityLevel = cms.uint32(0)
    ),
```
So you see that `RunHPSValidationRealMuonsData.GenericTriggerSelection. hltInputTag` is set to `cms.InputTag("TriggerResults","","HLT")`. However this is running in step1 which is the `HLT` process and since the module is in a Path it is impossible for it to get the TriggerResults it wants.

**Comment by Dr15Jones** [comment]
From the 140.1 workflow
```
>>> print process.higPhotonJetHLTOfflineSource.dumpPython()
cms.EDAnalyzer("HigPhotonJetHLTOfflineSource",
    dirname = cms.untracked.string('HLT/Higgs/PhotonJet'),
    hltPathsToCheck = cms.vstring('HLT_Photon22_R9Id90_HE10_Iso40_EBOnly_PFMET40_v', 
        'HLT_Photon22_R9Id90_HE10_Iso40_EBOnly_VBF_v', 
        'HLT_Photon36_R9Id90_HE10_Iso40_EBOnly_PFMET40_v', 
        'HLT_Photon36_R9Id90_HE10_Iso40_EBOnly_VBF_v', 
        'HLT_Photon50_R9Id90_HE10_Iso40_EBOnly_PFMET40_v', 
        'HLT_Photon50_R9Id90_HE10_Iso40_EBOnly_VBF_v', 
        'HLT_Photon75_R9Id90_HE10_Iso40_EBOnly_PFMET40_v', 
        'HLT_Photon75_R9Id90_HE10_Iso40_EBOnly_VBF_v', 
        'HLT_Photon90_R9Id90_HE10_Iso40_EBOnly_PFMET40_v', 
        'HLT_Photon90_R9Id90_HE10_Iso40_EBOnly_VBF_v', 
        'HLT_Photon120_R9Id90_HE10_Iso40_EBOnly_PFMET40_v', 
        'HLT_Photon120_R9Id90_HE10_Iso40_EBOnly_VBF_v', 
        'HLT_Photon135_PFMET40_v', 
        'HLT_Photon135_VBF_v', 
        'HLT_Photon155_v', 
        'HLT_Photon250_NoHE_v', 
        'HLT_Photon300_NoHE_v'),
    hltProcessName = cms.string('HLT'),
    pfJetsToken = cms.InputTag("ak4PFJets"),
    pfMetToken = cms.InputTag("pfMet"),
    pfjetMinPt = cms.untracked.double(30.0),
    photonMinPt = cms.untracked.double(20.0),
    photonsToken = cms.InputTag("gedPhotons"),
    pvToken = cms.InputTag("offlinePrimaryVertices"),
    triggerAccept = cms.untracked.bool(True),
    triggerResultsToken = cms.InputTag("TriggerResults"),
    verbose = cms.untracked.bool(False)
)
```
Here the problem is `higPhotonJetHLTOfflineSource.triggerResultsToken` which is set to read the 'default' TriggerResults. However, with the upcoming framework change the default is the TriggerResults made in this process which is only accessible after all Paths have finished.

**Comment by Dr15Jones** [comment]
All of these problems stem from `dqmoffline_step` being a Path instead of an EndPath.

**Comment by Dr15Jones** [comment]
It looks like `dqmoffline_step` is sometimes a Path and sometimes an EndPath: see 
https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/Configuration/Applications/python/ConfigBuilder.py#L1944

It looks like this is the crude attempt to get the pick up the TriggerResults for the HLT process even when one uses an empty processName. However, the only module which uses an empty processName is `higPhotonJetHLTOfflineSource.triggerResultsToken` while all the other ones explicitly state 'HLT'. 
The second reason appears to be to avoid picking up the TriggerResults when we are running fast sim. I think that could best be handled by using the `Modifier` system.

**Comment by Dr15Jones** [comment]
Two of the three fixes are in #15809

**Comment by Dr15Jones** [comment]
If there are problems in existing configuration, those configurations will no longer run since we changed the warning message to be a fatal exception more than a year ago.
